tools/make_patch.c

```diff
@@ -0,0 +1,944 @@
+		void *next = *(void **)list;
+		free(list);
+		list = next;
+	}
+}
+
+void free_asmfiles(struct asmfile *asmfile)
+{
+	void *temp;
+	while (asmfile) {
+		if (asmfile->symbols) free_list(asmfile->symbols);
+		temp = asmfile;
+		asmfile = asmfile->next;
+		free(temp);
+	}
+}
+
+int create_symbol(struct symbol **tip, const int value, const char *name)
+{
+	struct symbol *symbol;
+
+	symbol = malloc(sizeof(struct symbol) + strlen(name) + 1);
+	if (!symbol) {
+		fprintf(stderr, "Error: Cannot allocate memory\n");
+		return -1;
+	}
+	symbol->value = value;
+	strcpy(symbol->name, name);
+
+	// Link it at the start of the rest of the list
+	symbol->next = *tip;
+	*tip = symbol;
+
+	return 0;
+}
+
+const struct symbol *find_symbol(const struct symbol *symbols, const char *name)
+{
+	const struct symbol *symbol = symbols;
+	int namelen = strlen(name);
+	while (symbol) {
+		int symbolnamelen = strlen(symbol->name);
+		if (namelen > symbolnamelen) {
+			symbol = symbol->next; 
+			continue;
+		}
+		// If we're looking for a local label, only compare the part starting from the period
+		char *compare = (char *)symbol->name;
+		if (name[0] == '.') compare += symbolnamelen - namelen;
+		if (strcmp(compare, name) == 0) return symbol;
+		symbol = symbol->next;
+	}
+
+	return NULL;
+}
+
+int parse_address_bank(char *buffer_input){
+	int bank;
+	char *buffer = malloc(strlen(buffer_input));
+	char *endptr_bank;
+	char *buffer_address;
+
+	strcpy(buffer, buffer_input);
+	buffer_address = strchr(buffer, ':');
+	if (buffer_address) *buffer_address++ = '\0';
+
+	// Verify it consists of expected elements
+	if (strlen(buffer) > 2 || !buffer_address || strlen(buffer_address) > 4) {
+		if (buffer_address) *--buffer_address = ':';
+		fprintf(stderr, "Error: Unexpected %s\n", buffer);
+		return -1;
+	}
+
+	// Parse the bank part
+	bank = strtol(buffer, &endptr_bank, 16);
+	if (endptr_bank != buffer + strlen(buffer)){
+		fprintf(stderr, "Error: Cannot parse bank %s:%s\n", buffer, buffer_address);
+		return -1;
+	}
+
+	return bank;
+}
+
+int parse_address_address(char *buffer_input){
+	int address;
+	char *buffer = malloc(strlen(buffer_input));
+	char *endptr_address;
+	char *buffer_address;
+
+	strcpy(buffer, buffer_input);
+	buffer_address = strchr(buffer, ':');
+	if (buffer_address) *buffer_address++ = '\0';
+
+	// Verify it consists of expected elements
+	if (strlen(buffer) > 2 || !buffer_address || strlen(buffer_address) > 4){
+		if (buffer_address) *--buffer_address = ':';
+		fprintf(stderr, "Error: Unexpected %s\n", buffer);
+		return -1;
+	}
+
+	// Parse the address part
+	address = strtol(buffer_address, &endptr_address, 16);
+	if (endptr_address != buffer_address + strlen(buffer_address)){
+		fprintf(stderr, "Error: Cannot parse bank %s:%s\n", buffer, buffer_address);
+		return -1;
+	}
+
+	return address;
+}
+
+int parse_abs_offset(int bank, int address, char type)
+{
+	int offset;
+
+	// Calculate the absolute offset in the ROM
+	if (type == 'w') {
```

Comment:
This would be more concise using a `switch` instead of many `if`/`else`s. (You can just return the value right away, no need for `int offset` to exist.)

Reply:
Resolved in https://github.com/pret/pokecrystal/pull/882/commits/2aed65736a099e6a2d80a2ac4c6d4a86ace4f27d